add option to select a leg when starting navigation

Zayankovsky opened the pull request:
### Description
Closes #4681. 
Maybe we also need to trigger `onNextRouteLegStart` callback in this method, but I'm not sure. 

### Changelog
<!--
Include changelog entry (e.g. Fixed an unexpected change in recenter button when resuming the app.).
See https://github.com/mapbox/navigation-sdks/blob/main/documentation/android-changelog-guidelines.md.
You can remove the changelog block and add a `skip changelog` label, when applicable.
 -->
```
<changelog>Added option to select a leg when starting navigation</changelog>
```

alexander-kulikovskii commented on libnavigation-core/src/test/java/com/mapbox/navigation/core/MapboxNavigationTest.kt:
```diff
@@ -609,6 +609,20 @@ class MapboxNavigationTest {
+        assertTrue(mapboxNavigation.updateLegIndex(2))
+    }
+
+    @Test
+    fun `updateLegIndex should return false when route leg is not updated`() {
+        every { tripSession.updateLegIndex(3) } returns false
+
+        assertFalse(mapboxNavigation.updateLegIndex(3))
+    }
```
```suggestion
    @Test
    fun `updateLegIndex should return false when set negative leg index`() {
        every { tripSession.updateLegIndex(-1) } returns false

        assertFalse(mapboxNavigation.updateLegIndex(-1))
    }
```

Zayankovsky replied:
I don't see how is that any different. There's no logic in `MapboxNavigation.updateLegIndex`, that depends on the sign of the `legIdex` argument.

Guardiola31337 commented:
> Maybe we also need to trigger `onNextRouteLegStart` callback in this method, but I'm not sure. 

We should probably take a closer look at this. @kmadsen as you worked on this, could you provide some more context here when you have chance? 🙏

Guardiola31337 commented:
Noting that Metalava is ❌ 

```
src/main/java/com/******/navigation/core/MapboxNavigation.kt:761: error: Added method com.******.navigation.core.MapboxNavigation.updateLegIndex(int) [AddedMethod]
Aborting: Found compatibility problems checking the public API (/root/code/libnavigation-core/src/main/java) against the API in /root/code/libnavigation-core/api/current.txt

> Task :libnavigation-core:checkApi FAILED
```

See [Metalava](https://github.com/mapbox/mapbox-navigation-android/blob/master/docs/metalava.md) docs.

Zayankovsky commented:
> Noting that Metalava is ❌
> 
> ```
> src/main/java/com/******/navigation/core/MapboxNavigation.kt:761: error: Added method com.******.navigation.core.MapboxNavigation.updateLegIndex(int) [AddedMethod]
> Aborting: Found compatibility problems checking the public API (/root/code/libnavigation-core/src/main/java) against the API in /root/code/libnavigation-core/api/current.txt
> 
> > Task :libnavigation-core:checkApi FAILED
> ```
> 
> See [Metalava](https://github.com/mapbox/mapbox-navigation-android/blob/master/docs/metalava.md) docs.

Fixed.

Guardiola31337 commented:
Noting that `unit-tests-core` are ❌ 

```
> Task :libnavigation-core:compileDebugUnitTestKotlin FAILED
e: /root/code/libnavigation-core/src/test/java/com/******/navigation/core/trip/session/MapboxTripSessionTest.kt: (565, 27): No value passed for parameter 'withTripService'
```

Rebasing from https://github.com/mapbox/mapbox-navigation-android/pull/4741 should do the trick.

kmadsen commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:
```diff
@@ -770,6 +770,16 @@ class MapboxNavigation(
+     * @param legIndex new leg index
+     * @return true if no errors occurred, false otherwise
+     */
+    fun updateLegIndex(legIndex: Int): Boolean {
+        return tripSession.updateLegIndex(legIndex)
+    }
```
https://github.com/mapbox/mapbox-navigation-android/pull/4739#issuecomment-904128945

Add a function to `arrivalProgressObserver` and the `onNextRouteLegStart` callback will be handled for you

```suggestion
    fun updateLegIndex(legIndex: Int): Boolean {
        return arrivalProgressObserver.updateLegIndex(legIndex: Int)
    }
```

Also, could update `updateLegIndex` to `navigateLegIndex` to make it fit in with `navigateNextRouteLeg`

kmadsen replied:
``` kotlin
interface ArrivalObserver {
    /**
     * Called when [MapboxNavigation.navigateNextRouteLeg] has been called and returns true,
     * this observer will be notified of the next route leg start.
     */
    fun onNextRouteLegStart(routeLegProgress: RouteLegProgress)
```

This comment won't really make sense after this too. Will you update the comment to
``` kotlin
     * Called when [MapboxNavigation.navigateNextRouteLeg] or [MapboxNavigation. updateLegIndex]
     * has been called and returns true, this observer will be notified of the next route leg start.

kmadsen replied:
Actually, I don't think we should add this updateLegIndex function. The request, is to start the navigation trip at an index https://github.com/mapbox/mapbox-navigation-android/issues/4681

So I think we should introduce a way to start the navigation trip at an index
https://github.com/mapbox/mapbox-navigation-android/pull/4739#issuecomment-904837527

kmadsen commented:
https://github.com/mapbox/mapbox-navigation-android/issues/4681#issuecomment-904830976

I'm wondering if we really want to start a directions route and then update the leg. In metrics, this could end up looking like a failure case. Because it started a route, and didn't arrive. 

It would be nice to be able to specify the index when the route is set. But we also need to handle alternatives. 
``` kotlin
    fun setRoutes(routes: List<DirectionsRoute>) {
```
But this will probably be updated to
```
    fun setRoutes(routes: List<DirectionsRoute>, routeIndex: Int, legIndex: Int) {
```

And that will be more consistent with what the navigator will have
```
public native Expected<String, RouteInfo> setRoute(@Nullable String routeResponse, int route, int leg, @NonNull ActiveGuidanceOptions options);
```

Zayankovsky commented:
> [#4681 (comment)](https://github.com/mapbox/mapbox-navigation-android/issues/4681#issuecomment-904830976)
> 
> I'm wondering if we really want to start a directions route and then update the leg. In metrics, this could end up looking like a failure case. Because it started a route, and didn't arrive.
> 
> It would be nice to be able to specify the index when the route is set. But we also need to handle alternatives.
> 
> ```kotlin
>     fun setRoutes(routes: List<DirectionsRoute>) {
> ```
> 
> But this will probably be updated to
> 
> ```kotlin
>     fun setRoutes(routes: List<DirectionsRoute>, routeIndex: Int, legIndex: Int) {
> ```
> 
> And that will be more consistent with what the navigator will have
> 
> ```java
> public native Expected<String, RouteInfo> setRoute(@Nullable String routeResponse, int route, int leg, @NonNull ActiveGuidanceOptions options);
> ```

I added optional `legIndex` parameter to `MapboxNavigation.setRoutes` function. Not sure about `routeIndex` parameter though. Are you suggesting the user should be able to make some non-first route primary?

Zayankovsky commented:
> Noting that `unit-tests-core` are ❌
> 
> ```
> > Task :libnavigation-core:compileDebugUnitTestKotlin FAILED
> e: /root/code/libnavigation-core/src/test/java/com/******/navigation/core/trip/session/MapboxTripSessionTest.kt: (565, 27): No value passed for parameter 'withTripService'
> ```
> 
> Rebasing from #4741 should do the trick.

Done.

Zayankovsky commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt:
```diff
@@ -29,16 +29,22 @@ internal class MapboxDirectionsSession(
         }
+        this.routes = routes
+        if (routes.isNotEmpty()) {
+            primaryRouteOptions = routes[0].routeOptions()
```
Well, I added an option to set a custom primary route index. I hope I understood you correctly.

LukasPaczos commented:
Introducing an explicit `route index` is a breaking change that will impact all of the consumers of the core module, both internally and externally.

I agree that the idea is worth discussing - it might be a better choice to use a two-dimensional variable to indicate which route is the primary, which was by far done by re-arranging the list. The capability was always there, just accessed differently, and we've discussed these options in the past.

If we'd like to introduce an explicit route index instead of using the route at index `0` as the primary route, we'd have to not only refactor the `RoutesObserver` interface to carry this information as well, but also refactor our other features to work seamlessly with this setup. For example, the route line APIs also work with an assumption that the route at index `0` is the primary route and we'd have to adjust those and other features as well.

I'd like to propose not introducing this change here. Let's focus on being able to specify the active leg of the route and extract the proposal to introduce a route index variable to a separate issue or PR since it will require more discussion and work to address well, and is not directly related to the issue being tackled.

Zayankovsky commented:
> Introducing an explicit `route index` is a breaking change that will impact all of the consumers of the core module, both internally and externally.
> 
> I agree that the idea is worth discussing - it might be a better choice to use a two-dimensional variable to indicate which route is the primary, which was by far done by re-arranging the list. The capability was always there, just accessed differently, and we've discussed these options in the past.
> 
> If we'd like to introduce an explicit route index instead of using the route at index `0` as the primary route, we'd have to not only refactor the `RoutesObserver` interface to carry this information as well, but also refactor our other features to work seamlessly with this setup. For example, the route line APIs also work with an assumption that the route at index `0` is the primary route and we'd have to adjust those and other features as well.
> 
> I'd like to propose not introducing this change here. Let's focus on being able to specify the active leg of the route and extract the proposal to introduce a route index variable to a separate issue or PR since it will require more discussion and work to address well, and is not directly related to the issue being tackled.

Actually I wasn't sure about this change as well, but I couldn't wait more for Kyle's reply. So I added this change anyway in a separate commit, so that it could be easily reverted, which is what I just did.

korshaknn commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:
```diff
@@ -515,12 +515,14 @@ class MapboxNavigation(
      * Use [RoutesObserver] and [MapboxNavigation.registerRoutesObserver] to observe whenever the routes list reference managed by the SDK changes, regardless of a source.
      *
      * @param routes a list of [DirectionsRoute]s
+     * @param legIndex starting leg to follow
```
```suggestion
     * @param legIndex starting leg to follow. By default 0 is passed and the first leg is used.
```
something like this

korshaknn commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:
```diff
@@ -515,12 +515,14 @@ class MapboxNavigation(
      */
-    fun setRoutes(routes: List<DirectionsRoute>) {
+    @JvmOverloads
+    fun setRoutes(routes: List<DirectionsRoute>, legIndex: Int = 0) {
```
```suggestion
    fun setRoutes(routes: List<DirectionsRoute>, initialLegIndex: Int = 0) {
```
can we replace everywhere `legIndex` with `initialLegIndex ` to make it more clear?

korshaknn commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt:
```diff
@@ -29,16 +29,22 @@ internal class MapboxDirectionsSession(
+    override var initialLegIndex = 0
+        private set
+
+    override fun setRoutes(routes: List<DirectionsRoute>, legIndex: Int) {
```
```suggestion
    override fun setRoutes(routes: List<DirectionsRoute>, initialLegIndex: Int) {
```

LukasPaczos reviewed (approved):
LGTM.

nit: we can remove the default leg index argument from the `MapboxNativeNavigator#setRoute` interface so that it's clear that it has to be provided now.

Zayankovsky commented:
> nit: we can remove the default leg index argument from the `MapboxNativeNavigator#setRoute` interface so that it's clear that it has to be provided now.

Removed.
